File: Flutter/Flutter/lib/pages/sat_pages/associarSat.dart
import 'dart:math';

import 'package:flutter/material.dart';
import 'package:flutter_gertec/services/globalValues.dart';
import 'package:flutter_gertec/services/operacaoSat.dart';
import 'package:flutter_gertec/services/retornoSat.dart';
import 'package:flutter_gertec/util/common_code.dart';
import 'package:flutter_gertec/widgets/widgetsgertec.dart';
import 'package:flutter_masked_text/flutter_masked_text.dart';

class PageAssociarSat extends StatefulWidget {
  @override
  _PageAssociarSatState createState() => _PageAssociarSatState();
}

class _PageAssociarSatState extends State<PageAssociarSat> {
  final cnpj = MaskedTextController(mask: "00.000.000/0000-00"); // CNPJ do contribuinte
  final cnpjSoftHouse = MaskedTextController(mask: "00.000.000/0000-00"); // CNPJ do Software House

  // Inicializa o código de ativação com um valor global, para o usuario não precisar ficar digitando
  final codigoAtivacao = TextEditingController(text: GlobalValues.codAtivarSat); // Codigo de Ativacao do Sat
  final assinaturaAc = TextEditingController(); // Assinatura Ac

  CommonGertec commonGertec = new CommonGertec(); //* Classe que possui partes de código comuns em diversas telas

  // Função para validar os valores digitos pelo usuario e realizar a ativação do SAT
  void associarSat() async {
    GlobalValues.codAtivarSat = codigoAtivacao.text; // Salva  o código de ativação para o usuario não precisar ficar digitando em todas as telas
    if (commonGertec.isCodigoValido(codigoAtivacao.text)) {
      if (assinaturaAc.text.length != 0) {
        if (commonGertec.removeMaskCnpj(cnpj.text).length != 14 || commonGertec.removeMaskCnpj(cnpjSoftHouse.text).length != 14) {
          WidgetsGertec.dialogoSat("Verifique o CNPJ digitado!", context: context);
        } else {
          //* Chama a função Invocar Operação Sat, que recebe como parâmetro a "operação invocada" e um "Map com as chaves e seus respectivos valores".
          RetornoSat retornoSat = await OperacaoSat.invocarOperacaoSat(
            // Passa como parâmetro um Map de argumentos de valores que deseja enviar
            args: {
              'funcao': "AssociarSAT",
              'random': Random().nextInt(999999),
              'codigoAtivar': codigoAtivacao.text.toString(),
              'cnpj': commonGertec.removeMaskCnpj(cnpj.text),
              'cnpjSoft': commonGertec.removeMaskCnpj(cnpjSoftHouse.text),
              'assinatura': assinaturaAc.text
            },
          );

          //? Caso deseje visualizar o retorno completo basta chamar [retornoSat.getResultadoCompleto]

          //* Está função [OperacaoSat.formataRetornoSat] recebe como parâmetro a operação realizada e um objeto do tipo RetornoSat
          //* Retorna uma String com os valores obtidos do retorno da Operação já formatados e prontos para serem exibidos na tela
          // Recomenda-se acessar a função e entender como ela funciona
          String retornoFormatado = OperacaoSat.formataRetornoSat(retornoSat: retornoSat);
          WidgetsGertec.dialogoSat(retornoFormatado, context: context);
        }
      } else {
        WidgetsGertec.dialogoSat("Assinatura AC Inválida!", context: context);
      }
    } else {
      WidgetsGertec.dialogoSat("Código de Ativação deve ter entre 8 a 32 caracteres!", context: context);
    }
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: SingleChildScrollView(
        child: Container(
          color: Colors.white,
          padding: EdgeInsets.only(top: 30),
          height: MediaQuery.of(context).size.height,
          width: MediaQuery.of(context).size.width,
          child: Column(
            crossAxisAlignment: CrossAxisAlignment.center,
            mainAxisSize: MainAxisSize.min,
            children: <Widget>[
              SizedBox(
                height: 30,
              ),
              WidgetsGertec.formField(cnpj, "CNPJ Contribuinte: ", textInputType: TextInputType.number),
              SizedBox(height: 20),
              WidgetsGertec.formField(cnpjSoftHouse, "CNPJ Software House: ", textInputType: TextInputType.number),
              SizedBox(height: 20),
              WidgetsGertec.formField(codigoAtivacao, "Código de Ativação: "),
              SizedBox(height: 20),
              WidgetsGertec.formField(assinaturaAc, "Assinatura AC: "),
              SizedBox(height: 20),
              SizedBox(
                width: 240,
                child: RaisedButton(
                  child: Text("Associar"),
                  onPressed: () {
                    associarSat();
                  },
                ),
              ),
            ],
          ),
        ),
      ),
    );
  }
}

File: React_Native/React_Native/src/pages/sat_pages/associarSat.js
import React, { Component } from "react";
import { Text, View, Alert,Picker,TextInput,Button, DeviceEventEmitter } from "react-native";
import { SafeAreaView } from "react-native-safe-area-context";
import {TextInputMask} from 'react-native-masked-text';

import ValidaCodigo from '../../components/ValidacaoCodigo';
import ValidacaoCnpj from "../../components/ValidacaoCnpj";

import OperacaoSat from '../../services/operacaoSat';
import RetornoSat from '../../services/retornoSat';


export default class AssociarSat extends Component{
    
    constructor(props) {
        super(props);    
        this.state = {
            codigoAtivacao: global.MyVar,
            assinaturaAc: '',
            cnpj: '',// CNPJ do contribuinte
            cnpjSH: '', // CNPJ do Software House
            
        };
    }
    
    dialogoSat(messageText){
        Alert.alert("Retorno", messageText);
    }

    associarSat(){
        // Salva  o código de ativação para o usuario não precisar ficar digitando em todas as telas 
        global.MyVar = this.state.codigoAtivacao

        // CRIA UMA INSTÂNCIAS DO COMPONENTE QUE VALIDA O CÓDIGO E CNPJ
        var validationCodigo = new ValidaCodigo();
        var validationCnpj = new ValidacaoCnpj();

        // Limpeza dos CNPJs
        var cnpj = validationCnpj.getClean(this.state.cnpj.toString());
        var cnpjSH = validationCnpj.getClean(this.state.cnpjSH.toString());

        if(validationCodigo.getValidation(this.state.codigoAtivacao)){
            if(this.state.assinaturaAc.length != 0 ){
                if(validationCnpj.getSize(cnpj) || validationCnpj.getSize(cnpjSH)){
                    this.dialogoSat("Verifique o CNPJ digitado!");
                }else{
                    //ENVIA PARA OPERAÇÃO SAT ONDE VAI CONECTAR COM O SAT DE ACORDO COM A FUNÇÃO
                    var operacaoSat = new OperacaoSat();
                    args = {'funcao': 'AssociarSAT', 
                            'random': Math.floor(Math.random() * 9999999).toString(),
                            'codigoAtivar': this.state.codigoAtivacao,
                            'cnpj': this.state.cnpj,
                            'cnpjSH': cnpjSH,
                            'assinatura': this.state.assinaturaAc
                    };
                    operacaoSat.invocarOperacaoSat(args);
            
                    DeviceEventEmitter.once("eventAssociar", event => { 
                        var retorno = event.associar;
                        var retornoSat = new RetornoSat(args['funcao'], retorno);
                        this.dialogoSat(operacaoSat.formataRetornoSat(retornoSat));
                    });
                }
            }else {
                dialogoSat("Assinatura AC Inválida!");
            }
        }else {
            dialogoSat("Código de Ativação deve ter entre 8 a 32 caracteres!");
        }
    }
    
    render() {
        return (
            <SafeAreaView>
                <View style={{ marginBottom:10 , marginTop: 20}}>
                        <Text style={{fontSize:20, fontWeight: 'bold'}}>CNPJ Contribuinte</Text>
                        <View 
                            style={{backgroundColor:'white',
                                width:300,
                                marginStart:30,
                                marginBottom:10}
                            }
                        >
                            <TextInputMask
                                type={'cnpj'}  
                                value={this.state.cnpj}
                                onChangeText={text => {
                                    this.setState({
                                    cnpj: text
                                    })
                                }}
                            />
                        </View>
                </View>
                    
                <View style={{ marginBottom:10, }}>
                    <Text style={{fontSize:20, fontWeight: 'bold'}}>CNPJ Software House</Text>
                    <View style={{backgroundColor:'white',
                            width:300,
                            marginStart:30,
                            marginBottom:10, }}>
                        <TextInputMask
                            type={'cnpj'}
                            value={this.state.cnpjSH}
                            onChangeText={text => {
                                this.setState({
                                cnpjSH: text
                                })
                            }}
                        />
                    </View>
                </View>

                <View style={{ marginBottom:10,flexDirection:'row',justifyContent:'space-between' }}>
                    <Text style={{fontSize:20, fontWeight: 'bold'}}>Código de Ativação</Text>

                    <TextInput           
                        style={{height: 50,backgroundColor: 'white', fontSize: 17,width:160, marginRight:10}}
                        value={this.state.codigoAtivacao}
                        onChangeText={(text) => this.setState({codigoAtivacao: text})}  
                    />
                </View>

                <View style={{ marginBottom:10,flexDirection:'row',justifyContent:'space-between' }}>
                    <Text style={{fontSize:20, fontWeight: 'bold'}}>Assinatura AC</Text>

                    <TextInput           
                        style={{height: 50,backgroundColor: 'white', fontSize: 17,width:200, marginRight:10}}  
                        value={this.state.assinaturaAc}
                        onChangeText={(text) => this.setState({assinaturaAc: text})}
                    />
                </View>
                    
                <View style={{marginTop: 30, marginStart: 20, marginEnd: 20, marginBottom: 2}}>
                    <Button
                        color= 'gray'
                        title="associar"
                        onPress={() => this.associarSat()}
                    />
                </View>
            </SafeAreaView>
        );
    }
}